src/lib/sqle/connection.service.ts

```diff
@@ -1,74 +1,223 @@
+
+  constructor(private _queryService: VantageQueryService, private _sessionService: VantageSessionService) {
+    const connectionState: IVantageConnectionState = this._getConnectionState();
+    if (connectionState && connectionState.username !== this._currentUsername) {
+      // mismatch, so clear
+      this._connections = [];
+      this._currentConnection = undefined;
+    }
+  }
+
+  private set _currentConnection(connection: ISQLEConnection) {
+    this._setConnectionState({
+      current: connection,
+      connections: this._connections,
+      username: this._currentUsername,
+    });
+    this._currentConnectionSubject.next(connection);
+  }
+  private get _currentConnection(): ISQLEConnection {
+    return this._currentConnectionSubject.getValue();
+  }
+  public get currentConnection(): ISQLEConnection {
+    return this._currentConnection;
+  }
+
+  private set _connections(connections: ISQLEConnection[]) {
+    this._setConnectionState({
+      current: this._currentConnection,
+      connections,
+      username: this._currentUsername,
+    });
+    this._connectionsSubject.next(connections);
+  }
+  private get _connections(): ISQLEConnection[] {
+    return this._connectionsSubject.getValue();
+  }
+  public get connections(): ISQLEConnection[] {
+    return this._connections;
+  }
+
+  public addConnectionAndConnect(connection: ISQLEConnection, options?: IConnectOptions): Observable<ISQLEConnection> {
+    if (this._getConnectionIndex(connection) > -1) {
+      throw Error('Connection already exists');
+    } else {
+      return this._pingAndSave(connection, true, options);
+    }
+  }
+
+  public addConnection(connection: ISQLEConnection, options?: IConnectOptions): Observable<ISQLEConnection> {
+    if (this._getConnectionIndex(connection) > -1) {
+      throw Error('Connection already exists');
+    } else {
+      return this._pingAndSave(connection, false, options);
+    }
+  }
 
-  public get current(): ISQLEConnection {
-    return current();
+  public connectToExistingConnection(
+    connection: ISQLEConnection,
+    options?: IConnectOptions,
+  ): Observable<ISQLEConnection> {
+    if (this._getConnectionIndex(connection) > -1) {
+      return this._pingAndSave(connection, true, options);
+    } else {
+      throw Error('Connection does not exist');
+    }
+  }
+  public removeConnection(connection: ISQLEConnection): ISQLEConnection {
+    const index: number = this._getConnectionIndex(connection);
+    if (index > -1) {
+      this._connections = [...this._connections.slice(0, index), ...this._connections.slice(index + 1)];
+      this._currentConnection =
+        this._currentConnection && this._connectionsAreEqual(this._currentConnection, connection)
+          ? undefined
+          : this._currentConnection;
+      return connection;
+    } else {
+      // connection does not exist but that is fine?
+      return undefined;
+    }
+  }
+
+  public disconnectFromCurrentConnection(): void {
```

Comment:
Im struggling to follow the `connects` and `disconnects`..

in my mind, `current` is just a pointer to the last one that was selected

`connections` has all the active connections.

With that in mind, then disconnecting checks the pointer, removes it if its the one you are trying to disconnect from and then removes it from the array.

Thoughts?

Reply:
So in this API, there is `disconnect` and `remove`. `disconnect` unsets the connection as current connection, but leaves it in the connections array. `remove` unsets the connection as current connection AND removes it from the connections array

Reply:
I guess my question is, why would we remove it from current and not from the array when disconnecting? 

current is just a pointer no?